File: internal/credentials/credentials.go
package credentials

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"strings"
)

type Credentials struct {
	Email    string `json:"email"`
	Password string `json:"password"`
}

func (c Credentials) Validate() error {
	if strings.TrimSpace(c.Email) == "" {
		return errors.New("missing email")
	}
	if strings.TrimSpace(c.Password) == "" {
		return errors.New("missing password")
	}

	return nil
}

func Load(email, password, passwordFile, credentialsFile string) (Credentials, error) {
	if credentialsFile != "" {
		creds, err := LoadFile(credentialsFile)
		if err != nil {
			return Credentials{}, err
		}
		return creds, creds.Validate()
	}

	if passwordFile != "" {
		content, err := os.ReadFile(passwordFile)
		if err != nil {
			return Credentials{}, fmt.Errorf("read password file: %w", err)
		}
		password = strings.TrimSpace(string(content))
	}

	if email == "" {
		email = os.Getenv("GRADESCOPE_EMAIL")
	}
	if password == "" {
		password = os.Getenv("GRADESCOPE_PASSWORD")
	}

	creds := Credentials{
		Email:    strings.TrimSpace(email),
		Password: strings.TrimSpace(password),
	}

	return creds, creds.Validate()
}

func LoadFile(path string) (Credentials, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return Credentials{}, fmt.Errorf("read credentials file: %w", err)
	}

	var creds Credentials
	if err := json.Unmarshal(content, &creds); err == nil {
		return creds, nil
	}

	for _, line := range strings.Split(string(content), "\n") {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}

		parts := strings.SplitN(line, "=", 2)
		if len(parts) != 2 {
			continue
		}

		key := strings.TrimSpace(parts[0])
		value := strings.Trim(strings.TrimSpace(parts[1]), `"'`)

		switch key {
		case "GRADESCOPE_EMAIL", "email":
			creds.Email = value
		case "GRADESCOPE_PASSWORD", "password":
			creds.Password = value
		}
	}

	if err := creds.Validate(); err != nil {
		return Credentials{}, fmt.Errorf("parse credentials file: %w", err)
	}

	return creds, nil
}

File: src/credentials.mjs
import fs from "node:fs/promises";
import { promptLine, promptSecret } from "./ui.mjs";

export async function loadCredentialFile(filePath) {
  const content = await fs.readFile(filePath, "utf8").catch((error) => {
    throw new Error(`read credentials file: ${error.message}`);
  });

  try {
    const parsed = JSON.parse(content);
    return {
      email: String(parsed.email || "").trim(),
      password: String(parsed.password || "").trim(),
    };
  } catch {
    const credentials = {
      email: "",
      password: "",
    };

    for (const rawLine of content.split("\n")) {
      const line = rawLine.trim();
      if (!line || line.startsWith("#")) {
        continue;
      }

      const parts = line.split("=");
      if (parts.length < 2) {
        continue;
      }

      const key = parts[0].trim();
      const value = parts.slice(1).join("=").trim().replace(/^["']|["']$/g, "");
      if (key === "email" || key === "GRADESCOPE_EMAIL") {
        credentials.email = value;
      }
      if (key === "password" || key === "GRADESCOPE_PASSWORD") {
        credentials.password = value;
      }
    }

    return credentials;
  }
}

export async function loadCredentials(options = {}, { promptForMissing = false } = {}) {
  let email = String(options.email || "").trim();
  let password = String(options.password || "").trim();

  if (options.credentialsFile) {
    const fileCredentials = await loadCredentialFile(options.credentialsFile);
    email = email || fileCredentials.email;
    password = password || fileCredentials.password;
  }

  if (!password && options.passwordFile) {
    password = (await fs.readFile(options.passwordFile, "utf8").catch((error) => {
      throw new Error(`read password file: ${error.message}`);
    })).trim();
  }

  email = email || String(process.env.GRADESCOPE_EMAIL || "").trim();
  password = password || String(process.env.GRADESCOPE_PASSWORD || "").trim();

  if (promptForMissing) {
    if (!email) {
      email = await promptLine("Gradescope email: ");
    }
    if (!password) {
      password = await promptSecret("Gradescope password: ");
    }
  }

  if (!email) {
    throw new Error("missing email");
  }
  if (!password) {
    throw new Error("missing password");
  }

  return { email, password };
}
